### PR TITLE
dmctl(dm): add config task-template

### PR DESCRIPTION
**Pull request by Ehco1996**
### What problem does this PR solve? 

Issue Number: ref #3583 

### What is changed and how it works?

* add new cmd `dmctl config task-template`
```
➜ dmctl config task-template  -h
show task template which is created by WebUI with task config format

Usage:
  dmctl config task-template [task-name] [flags]

Flags:
  -h, --help   help for task-template
```
* fix templates openapi path from `api/v1/task/templates` to  `api/v1/tasks/templates`

we don't **exporse this cmd to public user** , beascause it's just a workaround for user who user WebUI and want to get task config template that created by openapi  but want to dump with v1 format

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

Side effects

 - Possible performance regression

Related changes

 - Need to update the documentation

### Release note 

```release-note
`None`.
```

### CODE REVIEW

**Comment by Ehco1996** [comment]
will add dmctl it test after https://github.com/pingcap/tiflow/pull/3656 merged

File: dm/dm/ctl/master/config.go [hunk 1]
```diff
@@ -79,6 +80,25 @@ func newConfigTaskCmd() *cobra.Command {
 	return cmd
 }
 
+func newConfigTaskTemplateCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "template [task-name]",
```
Comment (Ehco1996) [inline dm/dm/ctl/master/config.go:85]: too hard to give this cmd a right name 😂  task-config? template? OpenAPITask(this is wrong)?

the usage scenarios is  user create task config in WebUI, but want to dump this config to file and operate this task config with dmctl, neew suggestions here

**Comment by lance6716** [comment]
it's `get-config task`, nothing to do with template

**Comment by Ehco1996** [comment]
> it's `get-config task`, nothing to do with template

but we already have `get-confing` task cmd, so we need another alias for this cmd , do you have some suggestions?

**Comment by lance6716** [comment]
> > it's `get-config task`, nothing to do with template
> 
> but we already have `get-confing` task cmd, so we need another alias for this cmd , do you have some suggestions?

the old `get-config task` can be replaced by this logic 🤔

**Comment by Ehco1996** [comment]
> the old `get-config task` can be replaced by this logic 🤔

you mean when user use`get-config task` we try get config from subtask in scheduler first, if  there is no task we fallback to  `openapi.Task` in etcd ?

I'm worried if there are two tasks with the same name: `test`, one in the scheduler and the other in etcd 😂

**Comment by lance6716** [comment]
subtask is generated by task. Currently `get-config task` will merge subtasks to task but some information is lost. If we can save original task I think it's always better. @GMHDBJD PTAL

**Comment by GMHDBJD** [comment]
can we use --print-sample-config like what we do in master/worker?
e.g.
```
config task --print-sample-config/template
```

**Comment by GMHDBJD** [comment]
Prefer `template <task|master|worker> <name>`

**Comment by Ehco1996** [comment]
/unhold


> Prefer `template <task|master|worker> <name>`

i finally choose to use 
```
dmctl config task-template  -h

show task template which is created by WebUI with task config format

Usage:
  dmctl config task-template [task-name] [flags]

Flags:
  -h, --help   help for task-template
```

beacuse there is no  need to support template of master and worker or source

File: dm/openapi/spec/dm.yaml [hunk 2]
```diff
@@ -731,7 +731,7 @@ paths:
             "application/json":
               schema:
                 $ref: "#/components/schemas/ErrorWithMessage"
-  /api/v1/task/templates:
+  /api/v1/tasks/templates:
```
Comment (Ehco1996) [inline dm/openapi/spec/dm.yaml:734]: this change is becasue i found that i miss addressed this [comment](https://github.com/pingcap/tiflow/pull/3656#discussion_r772212806) when i was writing it test

fortunately, this has not been released yet, so it does not matter if i make changes.

File: dm/tests/openapi/run.sh [hunk 3]
```diff
@@ -338,6 +338,53 @@ function test_multi_tasks() {
...
+		"\"result\": true" 1
+
+	diff $WORK_DIR/get_task_from_task.yaml $WORK_DIR/get_task_from_task_template.yaml || exit 1
+
+	# delete source success and clean data for other test
+	openapi_source_check "delete_source_with_force_success" "mysql-01"
+	openapi_source_check "delete_source_with_force_success" "mysql-02"
+	openapi_source_check "list_source_success" 0
+	openapi_task_check "get_task_list" 0
+	echo ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>TEST OPENAPI: MULTI TASK TEMPLATE"
```
Comment (Ehco1996) [inline dm/tests/openapi/run.sh:385]: fix typo

File: dm/dm/master/server.go [hunk 4]
```diff
@@ -2145,9 +2146,61 @@ func (s *Server) GetCfg(ctx context.Context, req *pb.GetCfgRequest) (*pb.GetCfgR
...
+			return resp2, nil
+		}
+		if task == nil {
+			resp2.Msg = "task not found"
+			// nolint:nilerr
+			return resp2, nil
+		}
+		toDBCfg := config.GetTargetDBCfgFromOpenAPITask(task)
+		if adjustDBErr := adjustTargetDB(ctx, toDBCfg); adjustDBErr != nil {
+			if err != nil {
```
Comment (lance6716) [inline dm/dm/master/server.go:2178]: do you mean `adjustDBErr`?

File: dm/dm/master/server.go [hunk 5]
```diff
@@ -2145,9 +2146,61 @@ func (s *Server) GetCfg(ctx context.Context, req *pb.GetCfgRequest) (*pb.GetCfgR
...
+		sourceCfgMap := make(map[string]*config.SourceConfig)
+		for _, cfg := range task.SourceConfig.SourceConf {
+			if sourceCfg := s.scheduler.GetSourceCfgByID(cfg.SourceName); sourceCfg != nil {
+				sourceCfgMap[cfg.SourceName] = sourceCfg
+			} else {
+				resp2.Msg = fmt.Sprintf("the source: %s of task not found", cfg.SourceName)
+				return resp2, nil
+			}
+		}
+		subTaskConfigList, err := config.OpenAPITaskToSubTaskConfigs(task, toDBCfg, sourceCfgMap)
```
Comment (lance6716) [inline dm/dm/master/server.go:2193]: I prefer we change the return type of `OpenAPITaskToSubTaskConfigs` to []*SubTaskConfig

File: dm/dm/master/server_test.go [hunk 6]
```diff
@@ -2031,16 +2032,29 @@ func (t *testMaster) TestGetCfg(c *check.C) {
...
+	openapiTask, err := fixtures.GenNoShardOpenAPITaskForTest()
+	openapiTask.Name = notExistTaskName
+	c.Assert(err, check.IsNil)
+	c.Assert(ha.PutOpenAPITaskTemplate(t.etcdTestCli, openapiTask, true), check.IsNil)
+	c.Assert(failpoint.Enable("github.com/pingcap/tiflow/dm/dm/master/MockSkipAdjustTargetDB", `return(true)`), check.IsNil)
+	resp2, err = server.GetCfg(context.Background(), &pb.GetCfgRequest{Name: notExistTaskName, Type: pb.CfgType_TaskTemplateType})
+	c.Assert(failpoint.Disable("github.com/pingcap/tiflow/dm/dm/master/MockSkipAdjustTargetDB"), check.IsNil)
+	c.Assert(err, check.IsNil)
+	c.Assert(resp2.Result, check.IsTrue)
+	c.Assert(strings.Contains(resp2.Cfg, fmt.Sprintf("name: %s", notExistTaskName)), check.IsTrue)
```
Comment (lance6716) [inline dm/dm/master/server_test.go:2056]: why it's named `notExistTaskName`, now we are expected it existing.

File: dm/tests/openapi/run.sh [hunk 7]
```diff
@@ -338,6 +338,53 @@ function test_multi_tasks() {
...
+	prepare_database
+
+	taskname="test-1"
+
+	# create and check source
+	openapi_source_check "create_source1_success"
+	openapi_source_check "create_source2_success"
+	openapi_source_check "list_source_success" 2
+
+	# crud task temoplate
```
Comment (lance6716) [inline dm/tests/openapi/run.sh:352]: this file has many typo

File: dm/dm/master/server.go (hunk 5 above)
Comment (Ehco1996) [reply dm/dm/master/server.go:2193]: add in https://github.com/pingcap/tiflow/pull/3991/commits/7be509f333f20075e9c167c61a02adb05926fa6e, and fix many typos (i opened spell checker for shell scripts now 😂)

File: dm/dm/master/server.go [hunk 8]
```diff
@@ -2145,9 +2146,57 @@ func (s *Server) GetCfg(ctx context.Context, req *pb.GetCfgRequest) (*pb.GetCfgR
...
+		}
+		if task == nil {
+			resp2.Msg = "task not found"
+			// nolint:nilerr
+			return resp2, nil
+		}
+		toDBCfg := config.GetTargetDBCfgFromOpenAPITask(task)
+		if adjustDBErr := adjustTargetDB(ctx, toDBCfg); adjustDBErr != nil {
+			if adjustDBErr != nil {
+				resp2.Msg = err.Error()
```
Comment (lance6716) [inline dm/dm/master/server.go:2179]:
```suggestion
				resp2.Msg = adjustDBErr.Error()
```

Comment (lance6716) [inline dm/dm/master/server.go:2179]:
```suggestion
		if adjustDBErr := adjustTargetDB(ctx, toDBCfg); adjustDBErr != nil {
				resp2.Msg = adjustDBErr.Error()
```

File: dm/dm/master/server_test.go [hunk 9]
```diff
@@ -2031,16 +2032,30 @@ func (t *testMaster) TestGetCfg(c *check.C) {
...
 	}
 	resp2, err := server.GetCfg(context.Background(), req2)
 	c.Assert(err, check.IsNil)
 	c.Assert(resp2.Result, check.IsFalse)
 	c.Assert(resp2.Msg, check.Equals, "task not found")
 
+	// test get task template
+	openapiTask, err := fixtures.GenNoShardOpenAPITaskForTest()
+	openapiTask.Name = notExistTaskName
+	c.Assert(err, check.IsNil)
```
Comment (lance6716) [inline dm/dm/master/server_test.go:2049]: we should check the returned err before using other return values. In some case, a returned nil will panic

File: dm/dm/master/server_test.go (hunk 6 above)
Comment (lance6716) [reply dm/dm/master/server_test.go:2057]: don't forget this comment

File: dm/tests/openapi/client/openapi_task_check [hunk 10]
```diff
@@ -295,6 +295,145 @@ def stop_task_success(task_name):
     print("stop_task_success")
 
 
+def create_task_template_success(task_name, tartget_table_name=""):
```
Comment (lance6716) [inline dm/tests/openapi/client/openapi_task_check:298]:
```suggestion
def create_task_template_success(task_name, target_table_name=""):
```

File: dm/tests/openapi/client/openapi_task_check [hunk 11]
```diff
@@ -295,6 +295,145 @@ def stop_task_success(task_name):
...
+    assert resp.status_code == 200
+    print("get_task_list resp=", data)
+    assert data["total"] == int(count)
+
+
+def import_task_template(success_count, failed_count):
+    url = API_ENDPOINT + "/templates/import"
+    resp = requests.post(url=url)
+    data = resp.json()
+    print("create_task_template_success resp=", data)
```
Comment (lance6716) [inline dm/tests/openapi/client/openapi_task_check:401]: import_task_template

please self review this file

File: dm/dm/master/server_test.go (hunk 6 above)
Comment (Ehco1996) [reply dm/dm/master/server_test.go]: addressed in https://github.com/pingcap/tiflow/pull/3991/commits/1be01edcb9609c10442f22146281d4c098ab83d6 and https://github.com/pingcap/tiflow/pull/3991/commits/a09177f635b30a2d9348c309083ef35a31a86452

File: dm/dm/master/server.go [hunk 12]
```diff
@@ -2147,9 +2148,57 @@ func (s *Server) GetCfg(ctx context.Context, req *pb.GetCfgRequest) (*pb.GetCfgR
 	if shouldRet {
 		return resp2, err2
 	}
-	// For the get-config command, you want to filter out fields that are not easily readable by humans,
-	// such as SSLXXBytes field in `Security` struct
+
+	formartTaskString := func(subCfgList []*config.SubTaskConfig) string {
+		sort.Slice(subCfgList, func(i, j int) bool {
```
Comment (lance6716) [inline dm/dm/master/server.go:2153]: we can move this sort inside `SubTaskConfigsToTaskConfig`

**Review by lance6716 — APPROVED** [review]

File: dm/dm/master/server.go (hunk 12 above)
Comment (Ehco1996) [reply dm/dm/master/server.go:2153]: yes ,but we still need sort for get task config, so left sort part in this func

**Comment by Ehco1996** [comment]
could you please take another look for this pr ? thanks 🧡🧡🧡

/cc @GMHDBJD @okJiang @WizardXiao

File: dm/tests/openapi/run.sh [hunk 13]
```diff
@@ -338,6 +338,53 @@ function test_multi_tasks() {
...
+		"config task-template $taskname --path $WORK_DIR/get_task_from_task_template.yaml" \
+		"\"result\": true" 1
+
+	diff $WORK_DIR/get_task_from_task.yaml $WORK_DIR/get_task_from_task_template.yaml || exit 1
+
+	# delete source success and clean data for other test
+	openapi_source_check "delete_source_with_force_success" "mysql-01"
+	openapi_source_check "delete_source_with_force_success" "mysql-02"
+	openapi_source_check "list_source_success" 0
+	openapi_task_check "get_task_list" 0
```
Comment (okJiang) [inline dm/tests/openapi/run.sh:384]: how about wrap these to one function

**Review by okJiang — APPROVED** [review]
LGTM

File: dm/tests/openapi/run.sh (hunk 13 above)
Comment (Ehco1996) [reply dm/tests/openapi/run.sh]: added in https://github.com/pingcap/tiflow/pull/3991/commits/d5e953012acb1368db958b9b44e52ba9ade3a236 and  https://github.com/pingcap/tiflow/pull/3991/commits/0f3dd429a027bdb542ab92587601d51997d56d36

**Pull request merged**